ui/src/image_annotator_polygon.ts

```diff
@@ -1,41 +1,113 @@
+      this.boundaryRect = undefined
+    }
+
     this.draggedPoint = null
     this.draggedShape = null
   }
 
   cancelAnnotating() {
     this.currPolygonPoints = []
+    this.boundaryRect = undefined
+  }
+
+  updateBoundaryRect({ x, y }: { x: U, y: U }) {
+    // TODO: Update boundary when increasing the x1, y1 or decreasing the x2, y2.
+    if (this.boundaryRect) {
+      if (x < this.boundaryRect.x1) this.boundaryRect.x1 = x
+      if (x > this.boundaryRect.x2) this.boundaryRect.x2 = x
+      if (y < this.boundaryRect.y1) this.boundaryRect.y1 = y
+      if (y > this.boundaryRect.y2) this.boundaryRect.y2 = y
+    } else this.boundaryRect = { x1: x, x2: x, y1: y, y2: y }
+  }
+
+  addCurrPolygonPoint({ x, y }: { x: U, y: U }) {
+    this.currPolygonPoints.push({ x, y })
+    this.updateBoundaryRect({ x, y })
+  }
+
+  removeLastPoint() {
+    // TODO: Update boundaries on remove.
+    this.currPolygonPoints.pop()
+  }
+
+  finishPolygon(tag: S) {
+    const
+      { x, y } = this.currPolygonPoints[0],
+      newPolygon = {
+        shape: {
+          polygon: {
+            vertices: [...this.currPolygonPoints],
+            boundaryRect: this.boundaryRect
+          }
+        },
+        tag
+      },
+      isPolygon = this.currPolygonPoints.length > 2
+    this.drawLine(x, y)
+    this.currPolygonPoints = []
+    this.boundaryRect = undefined
+    if (isPolygon) return newPolygon
   }
 
   onClick(cursor_x: U, cursor_y: U, color: S, tag: S): DrawnShape | undefined {
     if (!this.ctx) return
 
     this.ctx.beginPath()
     this.ctx.fillStyle = color
+    if (this.isIntersectingFirstPoint(cursor_x, cursor_y)) return this.finishPolygon(tag)
+    if (this.currPolygonPoints.length) this.drawLine(cursor_x, cursor_y)
+
+    this.addCurrPolygonPoint({ x: cursor_x, y: cursor_y })
+  }
 
-    if (this.isIntersectingFirstPoint(cursor_x, cursor_y)) {
-      const { x, y } = this.currPolygonPoints[0]
-      this.drawLine(x, y)
-      const newPolygon = { shape: { polygon: { vertices: [...this.currPolygonPoints] } }, tag }
-      this.currPolygonPoints = []
-      return newPolygon
+  move(dx: U, dy: U, movedShape?: DrawnShape) {
+    // Keep the polygon in the boundaries.
+    const movedPolygon = (this.draggedShape || movedShape)?.shape.polygon
+    const boundaryRect = movedPolygon?.boundaryRect
+    if (!movedPolygon || !boundaryRect) return
+
+    const
+      { width, height } = this.canvas,
+      { x1, x2, y1, y2 } = boundaryRect,
+      moveX = x1 + dx < 0 ? -x1 : x2 + dx > width ? width - x2 : dx,
+      moveY = y1 + dy < 0 ? -y1 : y2 + dy > height ? height - y2 : dy
+
+    if (moveX) {
+      boundaryRect.x1 = x1 + moveX
+      boundaryRect.x2 = x2 + moveX
+    }
+    if (moveY) {
+      boundaryRect.y1 = y1 + moveY
+      boundaryRect.y2 = y2 + moveY
     }
-    if (this.currPolygonPoints.length) this.drawLine(cursor_x, cursor_y)
 
-    this.currPolygonPoints.push({ x: cursor_x, y: cursor_y })
+    movedPolygon.vertices.forEach(p => {
```

Comment:
```suggestion
    if (moveX && moveY) movedPolygon.vertices.forEach(p => {
```

Reply:
This won't work. Did you mean `if(moveX || moveY)` ?

Reply:
Yep, you are right.